src/pkg/library/client/push.go

```diff
@@ -40,285 +35,114 @@ func UploadImage(filePath string, libraryRef string, libraryURL string) error {
-		}
-		err := apiSetTag(baseURL+"/v1/tags/"+containerID, imgTag)
+	if !found {
+		sylog.Verbosef("Image %s does not exist in library - creating it.\n", imageHash)
+		image, err = createImage(libraryURL, imageHash, container.GetID().Hex())
 		if err != nil {
 			return err
 		}
 	}
-	return nil
-}
-
-func apiCreate(o interface{}, url string) (id string, err error) {
-	s, err := json.Marshal(o)
-	if err != nil {
-		return "", fmt.Errorf("Error encoding object to JSON:\n\t%v", err)
-	}
-	req, err := http.NewRequest("POST", url, bytes.NewBuffer(s))
-	req.Header.Set("Content-Type", "application/json")
-
-	client := &http.Client{}
-	res, err := client.Do(req)
-	if err != nil {
-		return "", fmt.Errorf("Error making request to server:\n\t%v", err)
-	}
-	if res.StatusCode != http.StatusOK {
-		jRes, err := ParseErrorBody(res.Body)
-		if err != nil {
-			jRes = ParseErrorResponse(res)
-		}
-		return "", fmt.Errorf("Creation did not succeed: %d %s\n\t%v",
-			jRes.Error.Code, jRes.Error.Status, jRes.Error.Message)
-	}
-
-	// Decode the returned created object to find its ID
-	c := make(map[string]map[string]interface{})
-	err = json.NewDecoder(res.Body).Decode(&c)
-	if err != nil {
-		return "", fmt.Errorf("Error decoding ID from server response:\n\t%v", err)
-	}
-
-	return c["data"]["id"].(string), nil
-
-}
-
-func apiExists(url string) (id string, err error) {
-	res, err := http.Get(url)
-	if err != nil {
-		return "", fmt.Errorf("Error making request to server:\n\t%v", err)
-	}
-	if res.StatusCode == http.StatusOK {
-		c := make(map[string]map[string]interface{})
-		json.NewDecoder(res.Body).Decode(&c)
-		if err != nil {
-			return "", fmt.Errorf("Error decoding ID from server response:\n\t%v", err)
-		}
-		return c["data"]["id"].(string), nil
-	}
-	return "", nil
-}
 
-func apiGetTags(url string) (tags map[string]string, err error) {
-	res, err := http.Get(url)
-	if err != nil {
-		return nil, fmt.Errorf("Error making request to server:\n\t%v", err)
-	}
-	if res.StatusCode != http.StatusOK {
-		jRes, err := ParseErrorBody(res.Body)
+	if !image.Uploaded {
+		sylog.Infof("Now uploading %s to the library\n", filePath)
+		err = postFile(libraryURL, filePath, image.GetID().Hex())
 		if err != nil {
-			jRes = ParseErrorResponse(res)
+			return err
 		}
-		return nil, fmt.Errorf("Creation did not succeed: %d %s\n\t%v",
-			jRes.Error.Code, jRes.Error.Status, jRes.Error.Message)
-	}
-	c := make(map[string]map[string]string)
-	json.NewDecoder(res.Body).Decode(&c)
-	if err != nil {
-		return nil, fmt.Errorf("Error decoding ID from server response:\n\t%v", err)
+		sylog.Debugf("Upload completed OK\n")
+	} else {
+		sylog.Infof("Image is already present in the library - not uploading.\n")
 	}
-	return c["data"], nil
-
-}
 
-func apiSetTag(url string, t ImageTag) (err error) {
-	s, err := json.Marshal(t)
+	sylog.Debugf("Setting tags against uploaded image\n")
+	err = setTags(libraryURL, container.GetID().Hex(), image.GetID().Hex(), tags)
 	if err != nil {
```

Comment:
Could this be in a single line if err := ...; err != {... }

Reply:
Skipping this - it's kinda long in the assignment already.